src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,409 @@
+
+lemma edge_in_two_incidence_sets {i j : V} {e : sym2 V} (H_ne : i ≠ j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j → e = ⟦(i, j)⟧ :=
+begin
+  refine quotient.rec_on_subsingleton e (λ p, _),
+  rcases p with ⟨v, w⟩,
+  rw eq_iff,
+  rintros ⟨⟨_, H_i⟩, ⟨_, H_j⟩⟩,
+  cases (mem_iff.mp H_i) with H_i₁ H_i₂;
+  cases (mem_iff.mp H_j) with H_j₁ H_j₂,
+  { exfalso, apply H_ne, rw [H_i₁, H_j₁] }, -- i = v, j = v
+  { left, use [H_i₁.symm, H_j₂.symm] },     -- i = v, j = w
+  { right, use [H_j₁.symm, H_i₂.symm] },    -- i = w, j = v
+  { exfalso, apply H_ne, rw [H_i₂, H_j₂] }  -- i = w, j = w
+end
+
+lemma mem_incidence_sets_iff_eq {i j : V} {e : sym2 V} (h : G.adj i j) :
+  e ∈ G.incidence_set i ∧ e ∈ G.incidence_set j ↔ e = ⟦(i, j)⟧ :=
+begin
+  refine quotient.rec_on_subsingleton e (λ p, _),
+  rcases p with ⟨v, w⟩,
+  rw eq_iff,
+  simp only [incidence_set],
+  tidy,
+end
+
+lemma adj_iff_exists_edge_val {i j : V} : G.adj i j ↔ ∃ (e : G.edge_set), e.val = ⟦(i, j)⟧ :=
+by simp only [mem_edge_set, exists_prop, set_coe.exists, exists_eq_right, subtype.coe_mk]
+
+-- 1. ∑ e : E, M i e * M j e = 1, where i and j are adjacent.
+theorem adj_sum_of_prod_inc_one {i j : V} (H_adj : G.adj i j) :
+  ∑ (e : G.edge_set), G.inc_matrix R i e * G.inc_matrix R j e = (1 : R) :=
+begin
+  simp only [inc_matrix_apply, ite_prod_one_zero, sum_boole,
+             G.mem_incidence_sets_iff_eq H_adj, ←subtype.val_eq_coe],
+  rw adj_iff_exists_edge_val at H_adj,
+  rcases H_adj with ⟨e, H_e⟩,
+  simp only [←H_e, edge_val_equiv],
+  have H : filter (λ (x : G.edge_set), x = e) univ = {e},
+  { ext, simp only [true_and, mem_filter, mem_univ, mem_singleton] },
+  simp only [H, filter_congr_decidable, nat.cast_one, card_singleton]
+end
+
+-- 2. M i e * M j e = 0, where i, j distinct non-adjacent vertices, e an edge.
+theorem inc_matrix_prod_non_adj {i j : V} {e : G.edge_set} (Hne : i ≠ j) (H_non_adj : ¬ G.adj i j) :
+  G.inc_matrix R i e * G.inc_matrix R j e = 0 :=
+begin
+  by_cases H₁ : G.inc_matrix R i e = 0,
+  { rw [H₁, zero_mul] },
+  { rw [inc_matrix_not_zero, inc_matrix_one] at H₁,
+    by_cases H₂ : G.inc_matrix R j e = 0,
+    { rw [H₂, mul_zero] },
+    { rw [inc_matrix_not_zero, inc_matrix_one] at H₂,
+      exfalso,
+      apply H_non_adj,
+      rw [←mem_edge_set, ←G.edge_in_two_incidence_sets Hne ⟨H₁, H₂⟩],
+      exact G.edge_val_in_set } }
+end
+
+-- 3. (M i e) ^ 2 = M i e; with i a vertex, e an edge.
+@[simp]
+theorem inc_matrix_element_power_id {i : V} {e : G.edge_set} :
+  (G.inc_matrix R i e) * (G.inc_matrix R i e) = G.inc_matrix R i e :=
+by simp [inc_matrix_apply]
+
+-- 4. degree(i) = ∑ e : E, M i e; where i is a vertex.
+theorem degree_equals_sum_of_incidence_row {i : V} :
+  (G.degree i : R) = ∑ (e : G.edge_set), G.inc_matrix R i e :=
+begin
+  rw [inc_matrix_def, ←card_incidence_set_eq_degree],
+  simp only [sum_boole, nat.cast_inj, fintype.card_coe_filter (G.incidence_set_subset i)],
```

Comment:
I'm trying it on gitpod and really not sure why `finset.filter_mem_eq_inter` isn't working here, but I think it's better to make that work instead of making an overly specific lemma about that

Reply:
(You don't use `fintype.card_coe_filter` anywhere else)

Reply:
Ahh, I see what's going on. The filter is over `λ x : ↥G.edge_set`, which means it's actually `↑x ∈ blah`. I think a lemma like this is probably what's needed here:

```
filter (λ x : ↥set, p ↑x) other_set = filter (λ x : ambient_type, p x) set ∩ other_set ```

Reply:
I actually had a debate on that lemma on Zulip, and I got some help which I used here. (https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/Matrix.20on.20edge.20set.20for.20graphs) Do you mean I should focus on solving this particular lemma about the degree rather than use a helping lemma about filters on finsets?

Reply:
nono, sorry this is more about the helper lemma. I think the main content of your helper lemma is really overly specific, and it could probably be narrowed in scope a lot to turn that `filter` you want to get rid of in the proof into an intersection (which it what it basically is), and then work from there, because that'll be far more generally useful than a lemma talking about a specific-looking `filter` statement and only giving its cardinality

Reply:
also, just realised - any specific reason you're using `edge_set` instead of `edge_finset`? feel like that may be easier to work with considering everything is finite

Reply:
> also, just realised - any specific reason you're using `edge_set` instead of `edge_finset`? feel like that may be easier to work with considering everything is finite

I wanted to define the incidence matrix with G.edge_finset initially, but for some reason only the G.edge_set worked. I am not sure I had problems working with the edge set (instead of finset), but I might be wrong.

Reply:
> nono, sorry this is more about the helper lemma. I think the main content of your helper lemma is really overly specific, and it could probably be narrowed in scope a lot to turn that `filter` you want to get rid of in the proof into an intersection (which it what it basically is), and then work from there, because that'll be far more generally useful than a lemma talking about a specific-looking `filter` statement and only giving its cardinality

Will try to do it without that helper lemma then :smile: